python/GafferUI/MessageWidget.py

```diff
@@ -246,8 +234,1225 @@ def handle( self, level, context, msg ) :
+
+		# Sadly as QAbstractProxyModel is, well - abstract, we can't add a constructor of our own
+		# as python will complain we haven't called the base constructor. Uses _ to avoid mangling fun.
+		if not hasattr( self, "_colorMap" ) :
+			self._colorMap = {}
+			for l in _messageLevels() :
+				self._colorMap[ int(l) ] = GafferUI._Variant.toVariant( GafferUI._StyleSheet.styleColor( "foreground%s" % l ) )
+
+		return self._colorMap[ levelData ]
+
+	# Signal forwarding.
+
+	def __dataChanged( self, topLeft, bottomRight, roles ) :
+
+		self.dataChanged.emit( self.index( topLeft.row() * 2, 0 ), self.index( ( bottomRight.row() * 2 ) + 1, 0 ), roles )
+
+	def __modelReset( self ) :
+
+		self.modelReset.emit()
+
+	def __rowsAboutToBeInserted( self, parent, start, end ) :
+
+		self.rowsAboutToBeInserted.emit( QtCore.QModelIndex(), start * 2, end * 2 + 1 )
+
+	def __rowsInserted( self, parent, start, end ) :
+
+		self.rowsInserted.emit( QtCore.QModelIndex(), start * 2, end * 2 + 1 )
+
+	def __rowsAboutToBeRemoved( self, parent, start, end ) :
+
+		self.rowsAboutToBeRemoved.emit( QtCore.QModelIndex(), start * 2, end * 2 + 1 )
+
+	def __rowsRemoved( self, parent, start, end ) :
+
+		self.rowsRemoved.emit( QtCore.QModelIndex(), start * 2, end * 2 + 1 )
+
+
+# Provides filtering based on message level. This isn't directly used for search
+# filtering as we often want search not to affect the number messages displayed.
+class _MessageTableFilterModel( QtCore.QSortFilterProxyModel ) :
+
+	def __init__( self, level = IECore.MessageHandler.Level.Info, *kw ) :
+
+		QtCore.QSortFilterProxyModel.__init__( self, *kw )
+		self.setLevel( level )
+
+	def setLevel( self, level ) :
+
+		self.__maxLevel = level
+		self.invalidateFilter()
+
+	def getLevel( self ) :
+
+		return self.__maxLevel
+
+	# Overrides for methods inherited from QSortFilterProxyModel
+	# --------------------------------------------------------
+
+	def filterAcceptsRow( self, sourceRow, sourceParent ) :
+
+		levelIndex = self.sourceModel().index( sourceRow, _MessageTableModel.Column.Level, sourceParent )
+		return self.sourceModel().data( levelIndex, _MessageTableModel.ValueRole ) <= self.__maxLevel
+
+# The base TabelModel representing the underlying message data.
+class _MessageTableModel( QtCore.QAbstractTableModel ) :
+
+	ColumnCount = 3
+	Column = IECore.Enum.create( "Level", "Context", "Message" )
+
+	# A role to allow access the underlying Message data, without any display coercion.
+	ValueRole = 100
+
+	def __init__( self, messages = None, parent = None ) :
+
+		QtCore.QAbstractTableModel.__init__( self, parent )
+
+		self.__messages = messages
+
+	def setMessages( self, messages ) :
+
+		# We make use of existing rows here rather than resetting
+		# the model as it avoids flickering where the view first
+		# scrolls to the top, and then is re-scrolled back to the
+		# bottom.
+
+		firstDifference = messages.firstDifference( self.__messages ) if self.__messages else None
```

Comment:
`self.__messages` is false when we have an empty messages container, in which case we would think we had no differences. Should be `if self.__messages is not None`?

Reply:
Good spot, thanks! I think between us we've managed to get the fact that that line was almost 100% wrong. The default should have been `0` too! I've fixed it in place:

https://github.com/GafferHQ/gaffer/blob/a9a75df3660b88bbfe4940e23969ecf5e59d8a52/python/GafferUI/MessageWidget.py#L1367

Reply:
Teamwork!